src/dmd/expressionsem.d

```diff
@@ -7352,7 +7354,33 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (tb.ty == Tstruct)
             {
                 ad = (cast(TypeStruct)tb).sym;
-                semanticTypeInfo(sc, tb);
+
+                Identifier hook = global.params.tracegc ? Id._d_delstructTrace : Id._d_delstruct;
+                if (!verifyHookExist(exp.loc, *sc, Id._d_delstructImpl, "deleting struct with dtor", Id.object))
+                    return setError();
+
+                // Lower to .object._d_delstruct{,Trace}(exp.e1)
+                Expression id = new IdentifierExp(exp.loc, Id.empty);
+                id = new DotIdExp(exp.loc, id, Id.object);
+
+                auto tiargs = new Objects();
+                tiargs.push(exp.e1.type);
+                id = new DotTemplateInstanceExp(exp.loc, id, Id._d_delstructImpl, tiargs);
+                id = new DotIdExp(exp.loc, id, hook);
+
+                e = new CallExp(exp.loc, id, exp.e1);
+                /* Gag errors generated by calls to `_d_delstruct`, because they display
+                 * internal compiler information, which is unnecessary to the user.
+                 */
+                uint errors = global.startGagging();
+                e = e.expressionSemantic(sc);
+                global.endGagging(errors);
+
+                /* Return both the original DeleteExp and the lowered CallExp to
+                 * `_d_delstruct` in order to keep checking for dtor errors as
+                 * well as generate code for the call to `_d_delstruct`.
+                 */
+                e = Expression.combine(exp, e);
```

Comment:
Why do we still need the original expression and pass it all the way to e2ir? Can't we simply check here if the dtor is callable in the current context and output an error if not?

Reply:
I tried to do this, but certain errors are output from different contexts, which are difficult to replicate here.

Here's an example: this snippet [1] may be output an error if the dtor throws. In order to verify this, `canThrow` is called here [2]. `_d_delstruct(s)` is `nothrow` so that it's not mentioned in errors such as this one [3].

I agree that returning both expressions is inelegant, but I don't see a way to avoid the situation above without duplicating the logic from semantic3.d in expressionsem.d.

[1] https://github.com/dlang/dmd/blob/93108bb9ea6216d67fa97bb4842fb59f26f6bfc7/src/dmd/semantic3.d#L747-L750
[2] https://github.com/dlang/dmd/blob/93108bb9ea6216d67fa97bb4842fb59f26f6bfc7/src/dmd/blockexit.d#L113-L114
[3] https://github.com/dlang/dmd/blob/93108bb9ea6216d67fa97bb4842fb59f26f6bfc7/src/dmd/canthrow.d#L58-L64